File: core/src/io/github/neilljohnston/iwannajump/exmath/Smooth.scala
package io.github.neilljohnston.iwannajump.exmath

/**
  * Smoothing functions galore.
  * Huge thanks to Desmos Graphing Calculator for existing.
  */
trait Smooth {
    /**
      * Sin transformation, bungeeSin(0) = 0 and bungeeSin(1) = 1
      * @param x    A value
      * @return sin(x*pi/2)
      */
    def bungeeSin(x: Float): Float = math.sin(x * math.Pi / 2).toFloat

    /**
      * Linear interpolation.
      * @param a    Starting value
      * @param b    Ending value
      * @param t    Increment
      * @return t/1 between a and b
      */
    def lerp(a: Float, b: Float, t: Float): Float = (b - a) * t + a
}

File: core/src/io/github/neilljohnston/iwannajump/java/math/Smooth.java
package io.github.neilljohnston.iwannajump.java.math;

/**
 * Mathematical smoothing and interpolation functions.
 */
public class Smooth {
    /**
     * Bring x to a proportion of t between x and y.
     * 
     * @param x Starting number
     * @param y Target number
     * @param t Time
     * @return Interpolated value for x
     */
    public static float lerp(float x, float y, float t) {
        return x + (y - x) * t;
    }

    /**
     * Bring x to a proportion of t between x and y.
     *
     * @param x Starting number
     * @param y Target number
     * @param t Time
     * @return Interpolated value for x
     */
    public static double lerp(double x, double y, double t) {
        return x + (y - x) * t;
    }

    /**
     * A sin function with period of 2 and amplitude of 1.
     *
     * @param x A number
     * @return sin(one divided by two multiplied by pi multiplied by x)
     */
    public static float unitSin(float x) {
        return (float) Math.sin(Math.PI / 2 * x);
    }

    /**
     * A sin function with period of 2 and amplitude of 1.
     *
     * @param x A number
     * @return sin(one divided by two multiplied by pi multiplied by x)
     */
    public static double unitSin(double x) {
        return Math.sin(Math.PI / 2 * x);
    }
}
